x-pack/legacy/plugins/ml/public/data_frame_analytics/pages/analytics_exploration/components/exploration/use_explore_data.ts

```diff
@@ -0,0 +1,123 @@
+}
+
+export interface UseExploreDataReturnType {
+  errorMessage: string;
+  status: INDEX_STATUS;
+  tableItems: EsDoc[];
+}
+
+export const useExploreData = (
+  jobConfig: DataFrameAnalyticsOutlierConfig | undefined,
+  selectedFields: EsFieldName[],
+  setSelectedFields: React.Dispatch<React.SetStateAction<EsFieldName[]>>
+): UseExploreDataReturnType => {
+  const [errorMessage, setErrorMessage] = useState('');
+  const [status, setStatus] = useState(INDEX_STATUS.UNUSED);
+  const [tableItems, setTableItems] = useState([] as EsDoc[]);
+
+  useEffect(() => {
+    (async () => {
+      if (jobConfig !== undefined) {
+        setErrorMessage('');
+        setStatus(INDEX_STATUS.LOADING);
+
+        try {
+          const resultsField = jobConfig.dest.results_field;
+
+          const resp: SearchResponse<any> = await ml.esSearch({
+            index: jobConfig.dest.index,
+            size: SEARCH_SIZE,
+            body: {
+              query: { match_all: {} },
+              sort: [
+                {
+                  [`${resultsField}.outlier_score`]: {
+                    order: 'desc',
+                  },
+                },
+              ],
+            },
+          });
+
+          const docs = resp.hits.hits;
+
+          if (docs.length === 0) {
+            setTableItems([] as EsDoc[]);
+            setStatus(INDEX_STATUS.LOADED);
+            return;
+          }
+
+          if (selectedFields.length === 0) {
+            const newSelectedFields = getDefaultSelectableFields(docs, resultsField);
+            setSelectedFields(newSelectedFields);
+          }
+
+          // Create a version of the doc's source with flattened field names.
+          // This avoids confusion later on if a field name has dots in its name
+          // or is a nested fields when displaying it via EuiInMemoryTable.
+          const flattenedFields = getFlattenedFields(docs[0]._source, resultsField);
+          const transformedTableItems = docs.map(doc => {
+            const item = {} as Record<string, any>;
+            flattenedFields.forEach(ff => {
+              item[ff] = getNestedProperty(doc._source, ff);
+              if (item[ff] === undefined) {
+                // If the attribute is undefined, it means it was not a nested property
+                // but had dots in its actual name. This selects the property by its
+                // full name and assigns it to `item[ff]`.
+                item[ff] = doc._source[`"${ff}"`];
+              }
+              if (item[ff] === undefined) {
+                const parts = ff.split('.');
+                if (parts[0] === resultsField && parts.length >= 2) {
+                  parts.shift();
+                  item[ff] = doc._source[resultsField][parts.join('.')];
+                }
+              }
+            });
+            return {
+              ...doc,
+              _source: item,
+            };
+          });
+
+          setTableItems(transformedTableItems as EsDoc[]);
```

Comment:
it might be better to cast the object returned from the map (line 108) as an `EsDoc` rather casting it here when it is used.
Or even better, the return type of might able to be added to line 109 to avoid a cast altogether.